drivers/mlx90393/mlx90393.c

```diff
@@ -0,0 +1,625 @@
+ * General Public License v2.1. See the file LICENSE in the top level
+ * directory for more details.
+ */
+
+/**
+ * @ingroup     drivers_mlx90393
+ * @{
+ *
+ * @file
+ * @brief       Device driver implementation for the MLX90393
+ *
+ * @author      Michael Ristau <michael.ristau@fh-erfurt.de>
+ *
+ * @}
+ */
+
+#include <errno.h>
+#include "mlx90393.h"
+#include "mlx90393_constants.h"
+#include "mlx90393_params.h"
+#include "ztimer.h"
+#include "imath.h"
+
+#define ENABLE_DEBUG    0
+#include "debug.h"
+
+#if MODULE_MLX90393_SPI
+#define DEV_SPI             (dev->params->spi)
+#define DEV_CS_PIN          (dev->params->cs_pin)
+#define DEV_CLK             (dev->params->clk)
+#define SPI_MODE            (SPI_MODE_3)
+#elif MODULE_MLX90393_I2C
+#define DEV_I2C             (dev->params->i2c)
+#define DEV_ADDR            (dev->params->addr)
+#endif
+
+#define DEV_MODE            (dev->params->mode)
+#define DEV_INT_PIN         (dev->params->int_pin)
+#define DEV_ODR             (dev->params->odr)
+#define DEV_GAIN            (dev->params->gain)
+#define DEV_RESOLUTION      (dev->params->resolution)
+#define DEV_OSR_MAG         (dev->params->oversampling.mag)
+#define DEV_OSR_TEMP        (dev->params->oversampling.temp)
+#define DEV_DIG_FILT        (dev->params->dig_filt)
+#define DEV_REF_TEMP        (dev->ref_temp)
+#define CONN_TEST_DATA      (0xAF03)
+
+#define MLX90393_BM_READ_TIMEOUT    (10)
+
+/** Forward declaration of functions for internal use */
+static int _init_bus(const mlx90393_t *dev);
+static void _acquire(mlx90393_t *dev);
+static void _release(mlx90393_t *dev);
+static int _write_byte(mlx90393_t *dev, uint8_t data);
+static int _read_byte(mlx90393_t *dev, uint8_t *buffer);
+static int _write_bytes(mlx90393_t *dev, void *data, size_t len);
+static int _read_bytes(mlx90393_t *dev, void *buffer, size_t len);
+static int _check_status_byte(mlx90393_t *dev);
+static int _write_register(mlx90393_t *dev, uint8_t addr, uint16_t value);
+static int _read_register(mlx90393_t *dev, uint8_t addr, uint16_t *value);
+static int _write_register_bits(mlx90393_t *dev, uint8_t addr, uint16_t mask, uint16_t value);
+static int _calculate_temp(uint16_t raw_temp, uint16_t ref_temp);
+static int _get_gain_factor(mlx90393_gain_t gain);
+static int _reset(mlx90393_t *dev);
+static int _exit(mlx90393_t *dev);
+static int _is_avaiable(mlx90393_t *dev);
+static void _calculate_conv_time(mlx90393_t *dev);
+static int _read_measurement(mlx90393_t *dev, uint8_t *buffer);
+
+int mlx90393_init(mlx90393_t *dev, const mlx90393_params_t *params)
+{
+    assert(dev);
+    assert(params);
+
+    dev->params = params;
+
+    int error = 0;
+
+    if ((error = _init_bus(dev)) != 0) {
+        return error;
+    }
+
+    _acquire(dev);
+
+    /* exit all continuous measurement modes */
+    if ((error = _exit(dev)) != 0) {
+        _release(dev);
+        return error;
+    }
+    ztimer_sleep(ZTIMER_MSEC, MLX90393_COMMAND_EX_TIMEOUT);
```

Comment:
```suggestion
    ztimer_sleep(ZTIMER_MSEC, MLX90393_COMMAND_EX_TIMEOUT);

```
Sie sind soooooooooooooooooooooooooooo sparsam mit Zeilenschaltungen. Sie dürfen mittels Leerzeilen durchaus inhaltlich mehr strukturieren, hier und folgende Anmerkungen.